src/Components/WebAssembly/Server/src/DebugProxyLauncher.cs

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.CommandLineUtils;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Builder
+{
+    internal static class DebugProxyLauncher
+    {
+        private static readonly object LaunchLock = new object();
+        private static readonly TimeSpan DebugProxyLaunchTimeout = TimeSpan.FromSeconds(10);
+        private static Task<string> LaunchedDebugProxyUrl;
+
+        public static Task<string> EnsureLaunchedAndGetUrl(IServiceProvider serviceProvider)
+        {
+            lock (LaunchLock)
+            {
+                if (LaunchedDebugProxyUrl == null)
+                {
+                    LaunchedDebugProxyUrl = LaunchAndGetUrl(serviceProvider);
+                }
+
+                return LaunchedDebugProxyUrl;
+            }
+        }
+
+        private static async Task<string> LaunchAndGetUrl(IServiceProvider serviceProvider)
+        {
+            var tcs = new TaskCompletionSource<string>();
+
+            var environment = serviceProvider.GetRequiredService<IWebHostEnvironment>();
+            var executablePath = LocateDebugProxyExecutable(environment);
+            var muxerPath = DotNetMuxer.MuxerPathOrDefault();
+            var ownerPid = Process.GetCurrentProcess().Id;
+            var processStartInfo = new ProcessStartInfo
+            {
+                FileName = muxerPath,
+                Arguments = $"exec \"{executablePath}\" --owner-pid {ownerPid}",
+                UseShellExecute = false,
+                RedirectStandardOutput = true,
+            };
+            RemoveUnwantedEnvironmentVariables(processStartInfo.Environment);
+
+            var debugProxyProcess = Process.Start(processStartInfo);
+            var nowListeningRegex = new Regex("^\\s*Now listening on: (.*)$");
+            debugProxyProcess.OutputDataReceived += (sender, eventArgs) =>
+            {
+                var match = nowListeningRegex.Match(eventArgs.Data);
```

Comment:
We should add a timeout to the Regex for good measure since this is shipping product code.

Reply:
For this particular regex I'm not sure how it could get stuck doing anything expensive but have added a timeout now anyway.